openflow13/nx_action.go

```diff
@@ -1471,3 +1472,451 @@ func NewNXActionController(controllerID uint16) *NXActionController {
+}
+
+func (a *NXActionController2PropMeterId) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, a.Len())
+	var b []byte
+	n := 0
+
+	a.Length = a.PropHeader.Len() + 4
+	b, err = a.PropHeader.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	copy(data[n:], b)
+	n += int(a.PropHeader.Len())
+
+	binary.BigEndian.PutUint32(data[n:], a.MeterId)
+	return
+}
+
+func (a *NXActionController2PropMeterId) UnmarshalBinary(data []byte) error {
+	a.PropHeader = new(PropHeader)
+	n := 0
+
+	if err := a.PropHeader.UnmarshalBinary(data[n:]); err != nil {
+		return err
+	}
+	if len(data) < int(a.Length) {
+		return errors.New("the []byte is too short to unmarshal a full NXActionController2Prop* message")
+	}
+	n += int(a.PropHeader.Len())
+
+	a.MeterId = binary.BigEndian.Uint32(data[n:])
+	return nil
+}
+
+func NewMeterId(meterId uint32) *NXActionController2PropMeterId {
+	a := new(NXActionController2PropMeterId)
+	a.PropHeader = new(PropHeader)
+	a.Type = NXAC2PT_METER_ID
+	a.Length = a.PropHeader.Len() + 4
+	a.MeterId = meterId
+	return a
+}
+
+// Decode Controller2 Property types.
+func DecodeController2Prop(data []byte) (Property, error) {
+	t := binary.BigEndian.Uint16(data[:2])
+	var p Property
+	switch t {
+	case NXAC2PT_MAX_LEN:
+		p = new(NXActionController2PropMaxLen)
+	case NXAC2PT_CONTROLLER_ID:
+		p = new(NXActionController2PropControllerID)
+	case NXAC2PT_REASON:
+		p = new(NXActionController2PropReason)
+	case NXAC2PT_USERDATA:
+		p = new(NXActionController2PropUserdata)
+	case NXAC2PT_PAUSE:
+		p = new(NXActionController2PropPause)
+	case NXAC2PT_METER_ID:
+		p = new(NXActionController2PropMeterId)
+	}
+	err := p.UnmarshalBinary(data)
+	if err != nil {
+		return p, err
+	}
+	return p, nil
+}
+
+// NXActionController2 is NX action to output packet to the Controller set with a specified ID.
+type NXActionController2 struct {
+	*NXActionHeader
+	pad [6]uint8
+
+	props []Property
+}
+
+func (a *NXActionController2) Len() uint16 {
+	n := a.NXActionHeader.Len() + 6
+	for _, prop := range a.props {
+		n += prop.Len()
+	}
+	return n
+}
+
+func (a *NXActionController2) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, a.Len())
+	var b []byte
+	n := 0
+
+	a.Length = a.Len()
+	b, err = a.NXActionHeader.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	copy(data[n:], b)
+	n += int(a.NXActionHeader.Len())
+	n += 6
+
+	for _, prop := range a.props {
+		b, err = prop.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		copy(data[n:], b)
+		n += int(prop.Len())
+	}
+
+	return
+}
+
+func (a *NXActionController2) UnmarshalBinary(data []byte) error {
+	a.NXActionHeader = new(NXActionHeader)
+	n := 0
+
+	if err := a.NXActionHeader.UnmarshalBinary(data[n:]); err != nil {
+		return err
+	}
+	if len(data) < int(a.Length) {
+		return errors.New("the []byte is too short to unmarshal a full NXActionController2 message")
+	}
+	n += int(a.NXActionHeader.Len())
+	n += 6
+
+	for n < len(data) {
```

Comment:
Not prefer to use len(data) here, otherwise an error might be introduced if other actions is followed after this action. You can use the Length in the Header?

Reply:
Done.